core-library/src/main/java/org/silverpeas/core/contribution/content/form/DataRecordUtil.java

```diff
@@ -24,91 +24,106 @@
+    throw new IllegalAccessError("Utility class");
+  }
+
   /**
    * Updates the specified fields.
    */
-  static public void updateFields(String[] fieldNames,
-      DataRecord updatedRecord, DataRecord copiedRecord) throws FormException {
-    Field updatedField;
-    Field copiedField;
-
+  public static void updateFields(String[] fieldNames, DataRecord updatedRecord,
+      DataRecord copiedRecord, String language) throws FormException {
     for (String fieldName : fieldNames) {
-      updatedField = updatedRecord.getField(fieldName);
-      try {
-        copiedField = copiedRecord.getField(fieldName);
-        if (copiedField == null) {
-          continue;
-        }
-      } catch (FormException ignored) {
-        continue;
-      }
+      updateField(fieldName, updatedRecord, copiedRecord, language);
+    }
+  }
 
-      if (updatedField.getTypeName().equals(copiedField.getTypeName())) {
-        updatedField.setObjectValue(copiedField.getObjectValue());
-      } else {
-        updatedField.setValue(copiedField.getValue());
+  private static void updateField(String fieldName,
+    DataRecord updatedRecord, DataRecord copiedRecord, String language) throws FormException {
+    Field updatedField = updatedRecord.getField(fieldName);
+    try {
+      Field copiedField = copiedRecord.getField(fieldName);
+      if (copiedField != null) {
+        if (updatedField.getTypeName().equals(copiedField.getTypeName())) {
+          Object value = copiedField.getObjectValue();
+          if (value instanceof String) {
+            updatedField.setObjectValue(
+                applySubstitution((String) value, updatedRecord, language));
+          } else {
+            updatedField.setObjectValue(copiedField.getObjectValue());
+          }
+        } else {
+          updatedField.setValue(copiedField.getValue());
+        }
       }
+    } catch (FormException ignored) {
+      SilverLogger.getLogger(DataRecordUtil.class)
+          .warn("Field '" + fieldName + "' is unknown", ignored);
     }
   }
 
   /**
-   * Returns : "But who is xoxox ?" for : "But who is ${foo}" when xoxox is the value of the foo
+   * Returns : "But who is xoxox ?" for : "But who is ${foo} ?" when xoxox is the value of the foo
    * field, The resolvedVar is used to detect recursive call like : foo = "${foo}"
    */
-  static public String applySubstitution(String text, DataRecord data,
-      String lang) {
-    return applySubstitution(text, data, lang, new ArrayList<>());
+  public static String applySubstitution(String text, DataRecord data, String lang) {
+    String appliedText = applySubstitution(text, data, lang, new ArrayList<>());
+    // applying twice to replace variables in substituted text
+    appliedText = applySubstitution(appliedText, data, lang, new ArrayList<>());
+    return appliedText;
   }
 
-  static private String applySubstitution(String text, DataRecord data,
+  private static String applySubstitution(String text, DataRecord data,
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://integration.silverpeas.org/sonar/coding_rules#rule_key=squid%3AS3776)